### ISSUE TITLE
about l2 regularization

### ISSUE THREAD

**Issue by graytowne** [geffy/tffm#5]
hello geffy,
tffm is very useful, here I have a small question:
how do you deal with the l2 regularization when input is sparse, I mean I look at your code, your regularization term include all of the parameters, but when the input data is very sparse, factorization machine only use a small part of parameters(e.g. not all of the w at order 1 can be used), if you do like this, all of the parameters will change, instead of only changing those part of parameters. Do you think it is a issue?

**Comment by geffy** [comment]
I believe you are right.

If you use tffm with batch_size=-1 even with sparse data -- it's ok to have such l2 regularization.
But if your data is very sparse and batch_size is small -- yes, current implementation can produce issue.

I think it is possible to implement l2-penalty only for factors used in current batch, but it can hurt performance. 

Add this to my todos.

**Comment by graytowne** [comment]
yes, you are right, because you have to index those parameters that have to be updated, and indexing in TensorFlow is inefficient.

**Comment by HoriuchiTakashi** [comment]
FTRL(Follow-the-Regularized-Leader) algorithm may work on very sparse situation.
http://www.jmlr.org/proceedings/papers/v15/mcmahan11b/mcmahan11b.pdf

And tensorflow has the optimizer.
https://www.tensorflow.org/versions/r0.10/api_docs/python/train.html#FtrlOptimizer

Maybe we can combine somehow?

**Comment by Vimos** [comment]
Even with large batch size, it seems that the loss does not converge, the best performance is achieved with full batch.

**Comment by geffy** [comment]
Finally I implement another approach for regularization -- it isn't faster than previous, but penalize only factors actually used in a batch. It use weights instead of indexing.

Details can be seen here: https://github.com/geffy/tffm/blob/303fc4035db4c2ffe401c9ccb54894e4db2e7085/tffm/core.py#L180-L189

Seems like this resolves the original issue.
@graytowne, are you agree?
